Remove all instances of low-level DH use in libssl

mattcaswell opened the pull request:
To prepare for the forthcoming deprecation of all the remaining low-level DH functions (see #13138), we remove all such usage from libssl. 

There are a few instances where a DH object is passed as part of the public API. In those cases we deprecate the functions/macros where that is the case and provider EVP_PKEY based replacements.

kroeckx commented on doc/man3/SSL_CTX_set_tmp_dh_callback.pod:
```diff
@@ -22,93 +32,73 @@ SSL_set_tmp_dh_callback, SSL_set_tmp_dh
+respectively. The parameters should be supplied in the I<dhpkey> argument as
+an B<EVP_PKEY> containg DH parameters. Ownership of the I<dhpkey> value is
+passed to the B<SSL_CTX> or B<SSL> object as a result of this call, and so the
+caller should not free it if the function call is succesful.
```
I think the text should discourage the use of those functions, and that they should use the built-in parameters instead.

mattcaswell replied:
I've strengthed the discouragement of this.

kroeckx commented on test/sslapitest.c:
```diff
@@ -8045,7 +8046,193 @@ static int test_ssl_dup(void)
+ * Test  7: Explicit SSL_CTX auto off, custom DH params via DH
+ * Test  8: Explicit SSL auto off, custom DH params via DH
+ * Test  9: Explicit SSL_CTX auto off, custom DH params via callback
+ * Test 10: Explicit SSL auto off, custom DH params via callback
```
I would like to see some tests that confirm that we select the correct DH parameters depending on the certificate/cipher.

mattcaswell replied:
I've added a new test in sslapitest.c for this.

levitte commented on apps/s_server.c:
```diff
@@ -3004,20 +3005,26 @@ static void print_connection_info(SSL *con)
+        = OSSL_DECODER_CTX_new_by_EVP_PKEY(&dhpkey, "PEM", "DH",
+                                           NULL, NULL);
+    if (decoderctx != NULL)
+        OSSL_DECODER_from_bio(decoderctx, bio);
```
There is now a `load_keyparams` in libapps...  does it still make sense to have a specialized `load_dh_params`?

mattcaswell replied:
I've made `load_dh_params` a simple wrapper around `load_keyparams` which additionally confirms that we have DH parameters.

levitte commented on include/openssl/ssl.h.in:
```diff
@@ -2130,12 +2141,14 @@ void SSL_set_default_read_buffer_len(SSL *s, size_t len);
+DEPRECATEDIN_3_0(void SSL_set_tmp_dh_callback(SSL *ssl,
+                                              DH *(*dh) (SSL *ssl,
+                                                         int is_export,
+                                                         int keylength)))
```
``` C
#  ifndef OPENSSL_NO_DEPRECATED_3_0
OSSL_DEPRECATEDIN_3_0
void SSL_CTX_set_tmp_dh_callback(SSL_CTX *ctx,
                                 DH *(*dh) (SSL *ssl, int is_export,
                                            int keylength));
OSSL_DEPRECATEDIN_3_0
void SSL_set_tmp_dh_callback(SSL *ssl,
                             DH *(*dh) (SSL *ssl, int is_export,
                                        int keylength));
#  endif
```

mattcaswell replied:
Fixed.

levitte commented on ssl/statem/statem_clnt.c:
```diff
@@ -3079,44 +3067,48 @@ static int tls_construct_cke_dhe(SSL *s, WPACKET *pkt)
-    prime_len = BN_num_bytes(DH_get0_p(dh_clnt));
+
+    /* Generate encoding of server key */
+    encoded_pub_len = EVP_PKEY_get1_tls_encodedpoint(ckey, &encoded_pub);
```
Travis complains about `EVP_PKEY_get1_tls_encodedpoint`...  missing an include file or has it been renamed by now?

https://travis-ci.com/github/openssl/openssl/jobs/405072122#L3847-L3850

mattcaswell replied:
This was the result of a merge error when I rebased. Fixed now.

mattcaswell commented:
All comments addressed. I've rebased and pushed the updates.

The new test reveals that if we use a ciphersuite without a cert and select a 128-bit symmetric cipher, then we end up using 1024-bit DH params. This seems very small.

t8m commented:
> All comments addressed. I've rebased and pushed the updates.
> 
> The new test reveals that if we use a ciphersuite without a cert and select a 128-bit symmetric cipher, then we end up using 1024-bit DH params. This seems very small.

Yep, we should use at least 2048 bit DH params unless 1024 bit cert is used.

mattcaswell commented:
Fix ups pushed addressing a mem leak reported by travis. I also made a tweak to the documentation.

kroeckx commented:
> And what about 256 bit symmetric ciphers?

Note that it's not because it's using AES256 that it reaches the
256 bit security level.

I currently see:
openssl ciphers -s -v DEFAULT:@SECLEVEL=5
TLS_AES_256_GCM_SHA384         TLSv1.3 Kx=any      Au=any   Enc=AESGCM(256) Mac=AEAD
TLS_CHACHA20_POLY1305_SHA256   TLSv1.3 Kx=any      Au=any   Enc=CHACHA20/POLY1305(256) Mac=AEAD
ECDHE-ECDSA-AES256-GCM-SHA384  TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AESGCM(256) Mac=AEAD
ECDHE-RSA-AES256-GCM-SHA384    TLSv1.2 Kx=ECDH     Au=RSA   Enc=AESGCM(256) Mac=AEAD
DHE-RSA-AES256-GCM-SHA384      TLSv1.2 Kx=DH       Au=RSA   Enc=AESGCM(256) Mac=AEAD
ECDHE-ECDSA-CHACHA20-POLY1305  TLSv1.2 Kx=ECDH     Au=ECDSA Enc=CHACHA20/POLY1305(256) Mac=AEAD
ECDHE-RSA-CHACHA20-POLY1305    TLSv1.2 Kx=ECDH     Au=RSA   Enc=CHACHA20/POLY1305(256) Mac=AEAD
DHE-RSA-CHACHA20-POLY1305      TLSv1.2 Kx=DH       Au=RSA   Enc=CHACHA20/POLY1305(256) Mac=AEAD
ECDHE-ECDSA-AES256-SHA384      TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AES(256)  Mac=SHA384
ECDHE-RSA-AES256-SHA384        TLSv1.2 Kx=ECDH     Au=RSA   Enc=AES(256)  Mac=SHA384
DHE-RSA-AES256-SHA256          TLSv1.2 Kx=DH       Au=RSA   Enc=AES(256)  Mac=SHA256

And I'm not sure all of them actually reach the 256 bit security
level that level 5 is supposed to reach.

In theory, if you want to reach the 256 bit security level, it
should be a 15360 key, which is really slow. We should not use
such a large parameters with the auto setting. We probably don't
even have such a large named group. My recommendation would be
to go for the 8192 bit key, reaching the 192 bit security level. I
think trying to reach anything higher than the 192 bit security level
using DHE doesn't make any sense.

You can actually argue that if the security level that's set allows
a 1024 (level 1) or 2048 bit (level 2) key, that can use it. But I
think that if we can reach the 128 bit security level (level 3) that
we should do that.

kroeckx commented:
> In theory, if you want to reach the 256 bit security level, it
> should be a 15360 key, which is really slow. We should not use
> such a large parameters with the auto setting. We probably don't
> even have such a large named group. My recommendation would be
> to go for the 8192 bit key, reaching the 192 bit security level. I
> think trying to reach anything higher than the 192 bit security level
> using DHE doesn't make any sense.

It seems we even don't support DH parameters larger then 10000.

t8m commented:
Let's be realistic. With non-anonymous ciphersuites (regardless of whether AES-128 or AES-256 is used) we would select 2k DH parameters if 2k RSA cert is used which is still the most common case. I do not think we need to try to be more strict than that with anonymous ciphersuites.

t8m commented:
And/or even better we could set the DH parameter size for anonymous ciphersuites based on the SECLEVEL setting with the low limit of 2048.

mattcaswell commented:
I've rebased this to resolve a conflict with master. Also fixed a travis issue and added a commit removing deprecated DH usage from ssltest_old (also resolving travis issues).

t8m commented on doc/man3/SSL_CTX_set_tmp_dh_callback.pod:
```diff
@@ -22,93 +32,77 @@ SSL_set_tmp_dh_callback, SSL_set_tmp_dh
+configure OpenSSL to use the default built-in DH parameters for the B<SSL_CTX>
+and B<SSL> objects respectively. Passing a value of 1 in the I<onoff> parameter
+switches the feature on, and passing a value of 0 switches it off. The default
+setting is off.
```
Hmm, should we make it on by default? I suppose that would be for another PR and OTC decision.

mattcaswell replied:
Yes, possibly we should. But, as you say, as another PR and pending an OTC decision.

t8m commented on ssl/s3_lib.c:
```diff
@@ -3776,22 +3768,21 @@ long ssl3_callback_ctrl(SSL *s, int cmd, void (*fp) (void))
 
     switch (cmd) {
-#ifndef OPENSSL_NO_DH
+#if !defined(OPENSSL_NO_DH ) && !defined(OPENSSL_NO_DEPRECATED_3_0)
```
Nit: extra space after OPENSSL_NO_DH

t8m commented on ssl/statem/statem_srvr.c:
```diff
@@ -2592,21 +2586,21 @@ int tls_construct_server_key_exchange(SSL *s, WPACKET *pkt)
-        DH_get0_pqg(dh, &r[0], NULL, &r[1]);
-        DH_get0_key(dh, &r[2], NULL);
+        /* These BIGNUMs need to be freed when we're finished */
+        freer = 1;
```
Instead of using this freer flag initialize the r array with NULLs. This is incorrect because theoretically any of the EVP_PKEY_get_bn_param could fail leaving the rest of the r uninitialized.

mattcaswell replied:
Err, they are already initialised to NULL earlier on in this function. We still need to use the `freer` variables, because other codepaths populate the r array with BIGNUMs that should *no*t be freed. However, this one populates it with values that *should* be freed.

t8m replied:
Ah, I overlooked the initialization in the unchanged code.

t8m reviewed (changes requested):
Some minor fixes needed.

t8m commented:
Also Travis failure is relevant.

mattcaswell commented on apps/s_server.c:
```diff
@@ -3003,20 +3004,16 @@ static void print_connection_info(SSL *con)
-    BIO_free(bio);
-    return ret;
+    if (dhpkey != NULL && !EVP_PKEY_is_a(dhpkey, "DH")) {
+        EVP_PKEY_free(dhpkey);
```
I've refactored this based on the commit from #13317 - so this PR is now dependent on that one.

mattcaswell commented:
> Also Travis failure is relevant.

Hopefully addressed. Lets see how it gets on now.

slontis commented on ssl/tls_depr.c:
```diff
@@ -143,5 +143,21 @@ HMAC_CTX *ssl_hmac_get0_HMAC_CTX(SSL_HMAC *ctx)
+EVP_PKEY *ssl_dh_to_pkey(DH *dh)
+{
+    EVP_PKEY *ret;
+    if (dh == NULL)
```
blankline

mattcaswell commented:
New fixup pushed addressing another travis issue. Also fixed the nit noted by @slontis.

mattcaswell commented:
Rebased now that #13317 has gone in.

Ping?

t8m commented:
You didn't push the rebase.

mattcaswell commented:
> You didn't push the rebase.

Hmm...the thing is...I did...its just not showing here for some reason.

mattcaswell commented:
Here's the branch in my personal github repo:

https://github.com/mattcaswell/openssl/commits/no-ssl-low-level-dh

It shows the head commit as being 0fbe6a3. But in this PR based on that branch, the head commit is still showing as c38a675.

WTF?

t8m commented:
Heh, closed with the intention to reopen but it is no longer possible. @mattcaswell perhaps you could reopen? If not, just open a new PR.

mattcaswell commented:
Nope, it won't let me reopen it either. Weird.

So, I opened a new PR instead: #13368
